substrate/codec/derive/src/encode.rs

```diff
@@ -0,0 +1,154 @@
+};
+
+type FieldsList = Punctuated<Field, Comma>;
+
+fn encode_fields<F>(
+	dest: &TokenStream,
+	fields: &FieldsList,
+	field_name: F,
+) -> TokenStream where
+	F: Fn(usize, &Option<Ident>) -> TokenStream,
+{
+	let recurse = fields.iter().enumerate().map(|(i, f)| {
+		let field = field_name(i, &f.ident);
+
+		quote_spanned! { f.span() =>
+			#dest.push(#field);
+		}
+	});
+
+	quote! {
+		#( #recurse )*
+	}
+}
+
+pub fn quote(data: &Data, type_name: &Ident, self_: &TokenStream, dest: &TokenStream) -> TokenStream {
+	let call_site = Span::call_site();
+	match *data {
+		Data::Struct(ref data) => match data.fields {
+			Fields::Named(ref fields) => encode_fields(
+				dest,
+				&fields.named,
+				|_, name| quote_spanned!(call_site => &#self_.#name),
+			),
+			Fields::Unnamed(ref fields) => encode_fields(
+				dest,
+				&fields.unnamed,
+				|i, _| {
+					let index = Index { index: i as u32, span: call_site };
+					quote_spanned!(call_site => &#self_.#index)
+				},
+			),
+			Fields::Unit => quote_spanned! { call_site =>
+				drop(#dest);
+			},
+		},
+		Data::Enum(ref data) => {
+			assert!(data.variants.len() < 256, "Currently only enums with less than 255 variants are encodable.");
+
+			let recurse = data.variants.iter().enumerate().map(|(i, f)| {
+				let name = &f.ident;
+				let index = super::index(f, i);
+
+				match f.fields {
+					Fields::Named(ref fields) => {
+						let field_name = |_, ident: &Option<Ident>| quote_spanned!(call_site => #ident);
+						let names = fields.named
+							.iter()
+							.enumerate()
+							.map(|(i, f)| field_name(i, &f.ident));
+
+						let encode_fields = encode_fields(
+							dest,
+							&fields.named,
+							|a, b| field_name(a, b),
+						);
+
+						quote_spanned! { f.span() =>
+							#type_name :: #name { #( ref #names, )* } => {
+								#dest.push_byte(#index as u8);
+								#encode_fields
+							}
+						}
+					},
+					Fields::Unnamed(ref fields) => {
+						let field_name = |i, _: &Option<Ident>| {
+							let data = stringify(i as u8);
+							let ident = from_utf8(&data).expect("We never go beyond ASCII");
+							let ident = Ident::new(ident, call_site);
+							quote_spanned!(call_site => #ident)
+						};
+						let names = fields.unnamed
+							.iter()
+							.enumerate()
+							.map(|(i, f)| field_name(i, &f.ident));
+
+						let encode_fields = encode_fields(
+							dest,
+							&fields.unnamed,
+							|a, b| field_name(a, b),
+						);
+
+						quote_spanned! { f.span() =>
+							#type_name :: #name ( #( ref #names, )* ) => {
+								#dest.push_byte(#index as u8);
+								#encode_fields
+							}
+						}
+					},
+					Fields::Unit => {
+						quote_spanned! { f.span() =>
+							#type_name :: #name => {
+								#dest.push_byte(#index as u8);
+							}
+						}
+					},
+				}
+			});
+
+			quote! {
+				match *#self_ {
+					#( #recurse )*,
+				}
+			}
+		},
+		// NOTE [ToDr] we currently don't use unions at all.
+		Data::Union(_) => unimplemented!(),
```

Comment:
I can't imagine how to implement unions at all. For example we have a type `union { a: A, b: B }` (where `A` and `B` are Encodable). To encode it we need to call either `A::encode` or `B::encode`, and there is no way to know which variant a particular union contains. To alleviate that, we could use some sort of tag which gives us essentially a `enum` : ) So I guess it's to panic with message like `unions are not supported` rather than `not yet implemented`.